app/models/reservation.rb

```diff
@@ -476,7 +477,40 @@ def name
 
   def update_overdue
     return true unless checked_out?
-    self.overdue = due_date < Time.zone.today
+    if due_date < Time.zone.today
+      self.overdue = true
+    else
+      # decrement the counter cache if we're changing overdue
+      # check for id to make sure this is not a new, unsaved reservation
+      # (only relevant in testing when overdue reservations are created)
+      equipment_model.decrement(:overdue_count) if overdue && id
```

Comment:
Just working this out "out loud":

1. If a reservation is checked out and flagged as overdue we increment the counter cache. This can happen through our rake task or through the `update_overdue` callback below.
2. If a reservation is overdue, we decrement the counter cache on the update that occurs during checkin.
3. If a reservation's overdue flag is toggled off in the `update_overdue` callback we also decrement the counter cache.

I think step 3 will have to be tweaked a bit after we merge in #1389 since the `update_overdue` callback will also run for reservations that have been checked in already (and therefore we might toggle overdue off and not want to decrement the counter cache).